core-bundle/src/Routing/Content/ArticleResolver.php

```diff
@@ -0,0 +1,46 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Routing\Content;
+
+use Contao\ArticleModel;
+use Contao\CoreBundle\Framework\ContaoFramework;
+use Contao\PageModel;
+
+class ArticleResolver implements ContentUrlResolverInterface
+{
+    public function __construct(private readonly ContaoFramework $framework)
+    {
+    }
+
+    public function resolve(object $content): ContentUrlResult
+    {
+        if (!$content instanceof ArticleModel) {
+            return ContentUrlResult::abstain();
```

Comment:
Why not return `null` instead of `new ContentUrlResult(null)`?

Reply:
Because returning a result object is more strict in my opinion.

Reply:
This does not make any sense to me, I'm afraid. I would prefer to return null. @contao/developers WDYT?

Reply:
> Here, we have more than 2.

What do you mean? We either have a result or not.

> `ContentUrlResult::abstain()` tells me exactly what I'm doing 

In what way? The naming does not make any sense at all. A voter can abstain from a vote, but a result cannot abstain. What exactly should an "abstained result" be? Do you mean "empty result"?

> it's way better DX in my opinion.

If that were such a great DX, why do we return `null` in `Model::findByPk()` if there is no match and not a `Model` object without data? And why does Symfony return `null` in `RequestStack::getCurrentRequest()` if there is no request instead of a `Request` object without data?

I have never seen such an API anywhere else.

Reply:
> What do you mean? We either have a result or not.

No, the way I understand it, is we have 4 results: https://github.com/contao/contao/pull/6596/files#diff-3c04b2eb4160f74d8b458c79520295bbd52a138a0e5926d5349baeac04c2fe12R19-R26

Well, I can only speak for myself. You asked what we think, so I'm telling you what I think:

I'm trying to understand this API design from the resolver's perspective. Because that's what I am going to write in the future. I will be the one having to implement the `ContentUrlResolverInterface` and returning the `ContentUrlResult` (or `null` in your proposal). Here's what my brain does:

1. "Hi, I am the `ContentUrlResolver`, I will now ask all the resolvers about my `$content`".
2. "Hello, `YanickResolver`, what would you like to do with `$content`?"
3. In your variant, I have to know that `null` means I abstain and the next resolver is asked. But this is not clear from just reading it. `null` could also mean "I want to generate an empty url". But maybe an empty URL is `''`, or maybe that is not allowed? I just don't know. But forcing me to return a `ContentUrlResult` means, I can only use one of the available methods and if this one is named `abstain()`, it's just very obvious to me what it means 😊 

But I think `ContentUrlResult::__construct()` should be `private`. I'll review that in a second.

Reply:
> A voter can abstain from a vote, but a result cannot abstain. What exactly should an "abstained result" be? Do you mean "empty result"?

It does not abstain from a result. The resolver abstains from resolving, exactly as a voter abstains from voting.

Reply:
Btw., this also prevents you from accidentially returning `NULL`, e.g. if a forward page is not found (as explained in https://github.com/contao/contao/pull/6596#discussion_r1452359462). A resolver **must decide to abstain**, not return NULL if anything else happened.

Reply:
> No, the way I understand it, is we have 4 results

That is not how I see it. The return type says `: ContentUrlResult`, so obviously, there is only one result. `ContentUrlResult::url()`, `ContentUrlResult::redirect()` and `ContentUrlResult::resolve()` are just three different factory methods.

> `null` could also mean "I want to generate an empty url".

If the return type is `ContentUrlResult|null`, I guess it is pretty obvious what `null` means → no result. If you want to return an empty URL, you would use `ContentUrlResult::url('')` because you are still returning something.

Reply:
see https://github.com/aschempp/contao/commit/a941ccd5f25836ab4949c0205c3a470605af3179 for a rename with clearer intentions.

Reply:
The new implementation makes even less sense than the previous one, I‘m afraid.

<img width="415" alt="" src="https://github.com/contao/contao/assets/1192057/af971999-d72d-4ab7-8896-22fe07674652">

IMHO it was correct that the resolver returned a result. It should just return `null` if it wants to abstain so we can replace `if ($result->isAbstained()` with `if (!$result)`.

Reply:
Or we add an `hasAbstained()` method to the resolver:

```php
foreach ($this->urlResolvers as $resolver) {
    $result = $resolver->resolve($contents[0]);

    if ($resolver->hasAbstained()) {
        continue;
    }
```

Although that is essentially the same as returning no result.

Reply:
I agree with @leofeyer here. I’d change it to `resolve(object $content): ResolverDecision|null` and make `ResolverDecision::$content` non-nullable.

To me the `ResolverDecision` in its current form does not look consistent. Calling `isRedirect()` or `hasTargetUrl()` or accessing `->content` would not make sense if `isAbstained() === true`, but that is not obvious from the outside. Consider for example some (constructed) use case like:

```php
if (!$resolver->resolve($content)->isRedirect()) {
	// Do something for regular (non-redirect) URLs…
}
```

This would probably do the wrong thing for the `isAbstained() === true` case. But if the return type of `resolve()` would be nullable, my IDE would tell me that I have to check that in advance, and the code would fail as it should.

If you decide that it has to stay like it is, I would at least make sure that all other methods of `ResolverDecision` throw an exception if `isAbstained() === true`. But to me this also shows that `null` would be a better fit here.

I’m fine with merging this either way but I’d vote for return type `ResolverDecision|null`.

Reply:
changed in 9b529db76eaae021b012e0b8b509871c33f6ec3d